### ISSUE TITLE
SAPT test suite failures on big-endian architectures

### ISSUE THREAD

**Issue by mbanck** [psi4/psi4#1462]
Psi4-1.2.1 on Debian shows several test suite failures on non-Intel hardware in the SAPT code that go like this:

```
90/140 Test #315: sapt1 ............................***Failed    7.40 sec
Traceback (most recent call last):
  File "/<<PKGBUILDDIR>>/builddir/stage/usr/bin/psi4", line 269, in <module>
    exec(content)
  File "<string>", line 53, in <module>
  File "/<<PKGBUILDDIR>>/builddir/stage/usr/lib/s390x-linux-gnu/psi4/driver/driver.py", line 492, in energy
    wfn = procedures['energy'][lowername](lowername, molecule=molecule, **kwargs)
  File "/<<PKGBUILDDIR>>/builddir/stage/usr/lib/s390x-linux-gnu/psi4/driver/procrouting/proc.py", line 3438, in run_sapt
    e_sapt = core.sapt(dimer_wfn, monomerA_wfn, monomerB_wfn)

RuntimeError: 
Fatal Error: Laplace Quadrature requested for (E_HUMO - E_LOMO)/(E_LUMO-E_HOMO) > 7.0 * 10^12, quadratures are not designed for this range.
Error occurred in file: /<<PKGBUILDDIR>>/psi4/src/psi4/lib3index/denominator.cc on line: 609
The most recent 5 function calls were:

psi::PsiException::PsiException(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char const*, int)

Exit Status: infile ( 1 ); autotest ( None ); sowreap ( None ); overall ( 1 )
```
See https://buildd.debian.org/status/fetch.php?pkg=psi4&arch=s390x&ver=1%3A1.2.1-2&stamp=1546705364&raw=0 for the full log.

Failing test cases from `quicktests` are `sapt1`, `sapt-dft1`, `tu5-sapt` and `pywrap-db1`, failing architectures are (so far) `s390x` and `powerpc`.

Anything I can do to investigate?

**Comment by loriab** [comment]
This came to mind for failing SAPT tests: https://github.com/psi4/psi4/pull/1398/commits/05a8eacd9d3a1b34e008f4078787dc6f4cb2c64f . Main action is in the SAPT iterator destructor. Fixed a double free. Sorry for the formatting mixed in.

But it doesn't look immediately relevant. Looks like someone intended to relax that error, https://github.com/psi4/psi4archive/blame/master/psi4/src/psi4/lib3index/denominator.cc#L197-L200.

My notes for other back-port candidates if 1.2.2 materializes: #1364, #1303, #1387

**Comment by mbanck** [comment]
I ran `sapt` on an s390x box now and compared the output to `output.ref` and one thing is weird:

```
-    Geometry (in Bohr), charge = 0, multiplicity = 1:
+    Geometry (in Bohr), charge = 0, multiplicity = 2:
```

Nothing else much stands out besides `output.ref` being from an older version so having some other strings here and there.

When SAPT would start

```
-        SAPT0  
-    Ed Hohenstein
-     6 June 2009
```
I immediately get the backtrace:
```
+Traceback (most recent call last):
+  File "/usr/bin/psi4", line 269, in <module>
+    exec(content)
+  File "<string>", line 53, in <module>
+  File "/usr/lib/s390x-linux-gnu/psi4/driver/driver.py", line 492, in energy
+    wfn = procedures['energy'][lowername](lowername, molecule=molecule, **kwargs)
+  File "/usr/lib/s390x-linux-gnu/psi4/driver/procrouting/proc.py", line 3438, in run_sapt
+    e_sapt = core.sapt(dimer_wfn, monomerA_wfn, monomerB_wfn)
+
+RuntimeError: 
+Fatal Error: Laplace Quadrature requested for (E_HUMO - E_LOMO)/(E_LUMO-E_HOMO) > 7.0 * 10^12, quadratures are not designed for this range.
```

I checked and adding `DENOMINATOR_ALGORITHM CHOLESKY` to `input.dat` makes the test pass:

```
        Nuclear Repulsion Energy..........................................PASSED
        SAPT0 Eelst.......................................................PASSED
        SAPT0 Eexch.......................................................PASSED
        SAPT0 Eind........................................................PASSED
        SAPT0 Edisp.......................................................PASSED
        SAPT0 Etotal......................................................PASSED
```

I'll try to step through the code first before I start recompiling with the patches you suggested.

**Comment by mbanck** [comment]
Ok so I think we are reading garbage from `PSIDATADIR/quadratures/1_x/R_avail.bin`:

On amd64 (my notebook):

```
588	    R_avail_file.read((char *)R_availp, nR * sizeof(double));
(gdb) 
590	    auto err_table = std::make_shared<Matrix>("Error Table (nR x nk)", nR, nk);
(gdb) p *R_availp
$8 = 2
```

And on s390x:

```
588         R_avail_file.read((char *)R_availp, nR * sizeof(double));
(gdb) 
590         auto err_table = std::make_shared<Matrix>("Error Table (nR x nk)", nR, nk);
(gdb) p *R_availp
$21 = 3.1620201333839779e-322
```

That also makes the for loop on line 602 exit immediately, resulting in the exception.

Is there anything odd with the way it is reading that?

**Comment by loriab** [comment]
@psi4/maintainers 

contents of those two `.bin` files are as follows. all the quadratures and the "tabelle" Rob got from the site he mentions in README, but I'm not sure how these two binary files got generated.

[R_avail file](https://github.com/psi4/psi4/blob/master/psi4/share/psi4/quadratures/1_x/R_avail.bin), 99 doubles read into array. not sure why not stored as ints.
```
0 2.000000
1 3.000000
2 4.000000
3 5.000000
4 6.000000
5 7.000000
6 8.000000
7 9.000000
8 10.000000
9 20.000000
10 30.000000
11 40.000000
12 50.000000
13 60.000000
14 70.000000
15 80.000000
16 90.000000
17 100.000000
18 200.000000
19 300.000000
20 400.000000
21 500.000000
22 600.000000
23 700.000000
24 800.000000
25 900.000000
26 1000.000000
27 2000.000000
28 3000.000000
29 4000.000000
30 5000.000000
31 6000.000000
32 7000.000000
33 8000.000000
34 9000.000000
35 10000.000000
36 20000.000000
37 30000.000000
38 40000.000000
39 50000.000000
40 60000.000000
41 70000.000000
42 80000.000000
43 90000.000000
44 100000.000000
45 200000.000000
46 300000.000000
47 400000.000000
48 500000.000000
49 600000.000000
50 700000.000000
51 800000.000000
52 900000.000000
53 1000000.000000
54 2000000.000000
55 3000000.000000
56 4000000.000000
57 5000000.000000
58 6000000.000000
59 7000000.000000
60 8000000.000000
61 9000000.000000
62 10000000.000000
63 20000000.000000
64 30000000.000000
65 40000000.000000
66 50000000.000000
67 70000000.000000
68 100000000.000000
69 200000000.000000
70 300000000.000000
71 400000000.000000
72 500000000.000000
73 700000000.000000
74 1000000000.000000
75 2000000000.000000
76 3000000000.000000
77 4000000000.000000
78 5000000000.000000
79 6000000000.000000
80 7000000000.000000
81 10000000000.000000
82 20000000000.000000
83 30000000000.000000
84 40000000000.000000
85 50000000000.000000
86 70000000000.000000
87 100000000000.000000
88 200000000000.000000
89 300000000000.000000
90 400000000000.000000
91 500000000000.000000
92 700000000000.000000
93 1000000000000.000000
94 2000000000000.000000
95 3000000000000.000000
96 4000000000000.000000
97 5000000000000.000000
98 7000000000000.000000
```
[error file](https://github.com/psi4/psi4/blob/master/psi4/share/psi4/quadratures/1_x/error.bin), 5247 doubles read into a Matrix. many zeros
```
0 0                       0.0212800000000000
0 1                       0.0435800000000000
0 2                       0.0596000000000000
0 3                       0.0707500000000000
0 4                       0.0782500000000000
0 5                       0.0828800000000000
0 6                       0.0851600000000000
0 7                       0.0855600000000000
0 8                       0.0855600000000000
0 9                       0.0000000000000000
0 10                       0.0000000000000000
0 11                       0.0000000000000000
0 12                       0.0000000000000000
...
```

[This block](https://github.com/psi4/psi4/blob/master/psi4/src/psi4/lib3index/denominator.cc#L566-L595) reads those bin files. It occurs three times in denominator.cc, once for each flavor of `*LaplaceDenominator.decompose()`, so refactoring into calling a free function is perhaps in order. After the arrays are filled, both are used to make sure the system can be fit w/i the availability and accuracy of the stored quadratures, https://github.com/psi4/psi4/blob/master/psi4/src/psi4/lib3index/denominator.cc#L196-L209. decompose happens once per SAPT calc, so perhaps ok to make text file? or alternate big-endian file? or better to detect host-endianness and read little-endian file accordingly? routine goes on to regex read a single quadrature file of <100 lines, so while this isn't time consuming compared to SAPT calc, the `.bin` files are larger than those in rest of dir.

**Comment by dgasmith** [comment]
Perhaps simpler to store as a numpy binary so endianness is handled automatically?

At first pass, it might simpler to simply check for endianess and flip if needed.

```
if (*(char *)&num == 1)
{
    printf("Little-Endian\n");
}
```

Not sure this catches all edge cases though. IIRC Motorola does something odd, according to a hacker news article that I may have read awhile ago...

**Comment by loriab** [comment]
Does #1464 fix those architectures?

**Comment by mbanck** [comment]
@loriab it does; though I kinda agree with @susilehtola that maybe the `.bin` files could be done with entirely.

Also, the binary patching in 2926990 won't work for the Debian package as it requires `git apply`, the usual Debian machinery just runs `patch` with preset options which can't handle it. So would have to wait for 1.2.2.

**Comment by loriab** [comment]
Yes, it seems elaborate just to keep metadata with the quadratures. And the quadratures themselves are not in the best state, considering we regex process them. And having thousands of them is a blot upon the `install` step.

They're downloaded from a website  (see [README](https://github.com/psi4/psi4/blob/master/psi4/share/psi4/quadratures/1_x/README)), so ideally both quadratures and metadata would be pulled into a json or similar library, then read into cpp. But this seemed a quick fix.

Will tag the PR for v1.2.2 backport. Ok to let those accumulate until all errors found and solved or do you need it sooner?

**Comment by susilehtola** [comment]
.. as well as remove the parsing steps.

**Comment by susilehtola** [comment]
Or, an even simpler fix would be to store the ```R_avail``` and ```error``` arrays as ASCII, like the quadrature rules themselves.

**Comment by loriab** [comment]
Then one would lose connection and `diff`-ability from the source/"upstream" website. After similar pain with updating physical constants, periodic table data, covalent radii, etc. we've stopped doing that and gone to the keep-original-format, write processor into py, write methods that can output json/python/c-header/etc setup of qcelemental.

I'm all for such a treatment of quadratures, as the existing setup isn't pretty and I resent those 2k files, too. It's just not in the timetable.

**Comment by susilehtola** [comment]
But the ASCII way is still possible :)

**Comment by mbanck** [comment]
I think storing in ASCII is a sensible thing, the data is big enough already and this will not add any significant storage overhead I think?

**Comment by loriab** [comment]
I also don't know exactly where the error array is coming from. The numbers in the README seem very few compared to the length of the error matrix.

The ASCII way is still possible. But Rob was no fool when it came to optimizing the right parts. It only gets run once per SAPT, so I don't see the block of a text file, but I'm ready to believe he had a reason to go binary.

**Comment by susilehtola** [comment]
> Then one would lose connection and `diff`-ability from the source/"upstream" website. After similar pain with updating physical constants, periodic table data, covalent radii, etc. we've stopped doing that and gone to the keep-original-format, write processor into py, write methods that can output json/python/c-header/etc setup of qcelemental.

Right. What I mean is just that in addition to the original files, the Psi4 source could have the parser that reads in and parses the original files to output C++ code. This functionality *already exists* in the SAPT code; the only thing necessary would be to take it out of Psi4 core, into a utility like the python file that saves the binary files, which would generate C++ sources for the tables.

**Comment by loriab** [comment]
Plan seems to be to go with one of these (https://xtensor.readthedocs.io/en/latest/api/io_index.html) when we adopt xtensor. Anyone really object to going with #1464 in the short term? I'm glad to approve an ascii workaround, but I don't have the time to re-rejigger when there's something that works.

**Comment by mbanck** [comment]
I don't mind either way, maybe just commit to the `1.2.x` branch and wait for a proper fix in `master`?; though I guess `v1.3` is supposed to be due soon as well?
